sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Redistribute.java

```diff
@@ -0,0 +1,156 @@
+    @Override
+    public PCollection<T> apply(PCollection<T> input) {
+      return input
+          .apply(ParDo.of(new PairWithBundleKey<T>()))
+          .apply(Reshuffle.<Integer, T>of())
+          .apply(Values.<T>create());
+    }
+  }
+
+  /**
+   * Pairs each element with a best-effort unique key.
+   *
+   * <p>We would like to efficiently pair each element in the input {@link PCollection} with a key
+   * that is as globally unique as possible, because reparallelization will be at most as good as
+   * the uniqueness of our generated keys.
+   *
+   * <p>An obvious way is to generate a new random number for each element. Then bundle "i" will
+   * generate a stream:
+   *
+   * <pre>
+   *   r_i0, r_i1, ...
+   * </pre>
+   *
+   * However, here we use a faster strategy that produces the same expected number of duplicates:
+   * bundle "i" generates a seed random number r_i, and then simply repeatedly increments it:
+   *
+   * <pre>
+   *   r_i, r_i + 1, r_i + 2, ...
+   * </pre>
+   *
+   * This seems less random, but here's why this produces the same number of duplicates:
+   *
+   * <p>Let us compute the expectation of how many times the number X will appear in the union of
+   * the sets of keys generated by each bundle - i.e. expected duplication ratio.
+   *
+   * <p>For any single number generated by any of the bundles, the answer is, with both strategies,
+   * 1/U, where U is the universe size - i.e. probability that this particular number equals X.
+   *
+   * <p>By <a href="https://en.wikipedia.org/wiki/Expected_value#Linearity">linearity of
+   * expectation</a>, for several bundles of total size N, the answer will be also the same - N/U.
+   *
+   * <p>Q.E.D.
+   */
+  private static class PairWithBundleKey<T> extends DoFn<T, KV<Integer, T>> {
+    private int currentBundleKey;
+
+    @Setup
+    public void setup() {
+      currentBundleKey = ThreadLocalRandom.current().nextInt();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      c.output(KV.of(currentBundleKey++, c.element()));
+    }
+  }
+
+  /** Implementation of {@link #byKey}. */
+  private static class ByKey<K, V>
+      extends PTransform<PCollection<KV<K, V>>, PCollection<KV<K, V>>> {
+    @Override
+    public PCollection<KV<K, V>> apply(PCollection<KV<K, V>> input) {
+      WindowingStrategy<?, ?> originalStrategy = input.getWindowingStrategy();
+      // If the input has already had its windows merged, then the GBK that performed the merge
+      // will have set originalStrategy.getWindowFn() to InvalidWindows, causing the GBK contained
+      // here to fail. Instead, we install a valid WindowFn that leaves all windows unchanged.
+      Window.Bound<KV<K, V>> rewindow =
+          Window.<KV<K, V>>into(
+                  new IdentityWindowFn<>(originalStrategy.getWindowFn().windowCoder()))
+              .triggering(new ReshuffleTrigger<>())
```

Comment:
Noting that `ReshuffleTrigger` is a bit of a hack and we need to get rid of it or come up with a way to incorporate it into the supported language of triggers.